lib/MusicBrainz/Server/Data/Utils.pm

```diff
@@ -770,6 +771,35 @@ sub contains_string {
     return any { $_ eq $string } @$array_ref;
 }
 
+# Given an array of MusicBrainz::Server::Entity::Alias objects and the UI
+# language, try to find a localized primary alias to display to the user.
+sub find_best_primary_alias {
+    my ($aliases_ref, $lang) = @_;
+
+    my $short_lang = substr($lang, 0, 2);
+    my ($best, $fallback);
+    foreach my $alias (@$aliases_ref) {
+        next if !defined $alias->locale || !$alias->primary_for_locale;
+
+        # If we find an exact match for the user's language, use it.
+        return $alias if $alias->locale eq $lang;
+
+        # Otherwise, favor more-generic aliases (e.g. "en" over "en_US").
+        if (substr($alias->locale, 0, 2) eq $short_lang &&
+            (!defined $best || length($alias->locale) == 2)) {
+            $best = $alias;
+        }
+        # If we find an English alias, use it as a fallback. This is likely
```

Comment:
I figured it'd be worthwhile falling back to English here since it looks like all of the UI languages that are presently listed use the Latin script.

Reply:
I think I mentioned it earlier, but a (future?) downside of this is that if e.g. a user selects Japanese as their UI language (I don't think they can at present), then they'll always see English aliases for Japanese artists, since I don't think there's any way to know that the artist entity's name is already in Japanese.

Reply:
You could in theory check if there's a Japanese artist alias that exactly matches the artist name (but you're right, if that data doesn't exist).

Reply:
Yeah, in practice I don't think that many editors add primary aliases that are identical to entity names. https://musicbrainz.org/artist/9ddd7abc-9e1b-471d-8031-583bc6bc8be9/aliases has a zillion aliases, but of the two matching the "Пётр Ильич Чайковский" entity name, one is a non-primary legal name for Russian and the other is a search hint. Oddly, there's a primary artist name Russian alias of "Пётр Чайковский".